src/github.com/matrix-org/dendrite/roomserver/types/types.go

```diff
@@ -119,3 +125,19 @@ type StateEntryList struct {
 	StateBlockNID StateBlockNID
 	StateEntries  []StateEntry
 }
+
+// A RoomRecentEventsUpdater is used to update the recent events in a room.
+// (On postgresql this wraps a database transaction that holds a "FOR UPDATE"
+//  lock on the a row holding the latest events for the room.)
+type RoomRecentEventsUpdater interface {
+	// Store the previous events referenced by an event.
+	StorePreviousEvents(eventNID EventNID, previousEventReferences []gomatrixserverlib.EventReference) error
+	// Check whether the eventReference is already referenced by another matrix event.
+	IsReferenced(eventReference gomatrixserverlib.EventReference) (bool, error)
+	// Set the list of latest events for the room.
+	SetLatestEvents(roomNID RoomNID, latest []StateAtEventAndReference) error
```

Comment:
Or would you like to see comments on how often SetLatest... and StorePrevious... are called?